embedded-graphics-core/src/rectangle/mod.rs

```diff
@@ -0,0 +1,760 @@
+    /// }
+    /// ```
+    ///
+    /// [`columns`]: #method.columns
+    /// [`points`]: https://docs.rs/embedded-graphics/latest/embedded_graphics/primitives/trait.Primitive.html#tymethod.points
+    pub fn rows(&self) -> Range<i32> {
+        self.top_left.y
+            ..self
+                .top_left
+                .y
+                .saturating_add(self.size.height.saturating_cast())
+    }
+
+    /// Returns the range of X coordinates in this rectangle.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use embedded_graphics::{prelude::*, primitives::Rectangle};
+    ///
+    /// let rect = Rectangle::new(Point::new(10, 20), Size::new(3, 4));
+    ///
+    /// assert_eq!(rect.columns(), 10..13);
+    /// ```
+    ///
+    /// By combining this method with [`rows`] it is possible to iterator over all pixels inside
+    /// the rectangle. This can be more flexible than using the [`points`] iterator, for example,
+    /// if a different iteration order is required or some operations should be called once per row.
+    ///
+    /// ```
+    /// use embedded_graphics::{prelude::*, primitives::Rectangle};
+    ///
+    /// let rect = Rectangle::new(Point::new(10, 20), Size::new(3, 4));
+    ///
+    /// // Iterate over all points starting from the top right corner and advancing downwards.
+    /// for x in rect.columns().rev() {
+    ///     for y in rect.rows() {
+    ///         // use x, y coordinates
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// [`rows`]: #method.rows
+    /// [`points`]: https://docs.rs/embedded-graphics/latest/embedded_graphics/primitives/trait.Primitive.html#tymethod.points
+    pub fn columns(&self) -> Range<i32> {
+        self.top_left.x
+            ..self
+                .top_left
+                .x
+                .saturating_add(self.size.width.saturating_cast())
+    }
+
+    /// Create a row-first, top-down iterator over all points in the rectangle.
+    pub fn points(&self) -> Points {
+        Points::new(self)
+    }
+
+    /// Returns `true` is the rectangle is zero sized.
+    ///
+    /// A rectangle is zero sized if the width or height are zero.
+    ///
+    /// # Examples
+    /// ```
+    /// use embedded_graphics::{prelude::*, primitives::Rectangle};
+    ///
+    /// let rect = Rectangle::new(Point::new(10, 20), Size::new(10, 20));
+    /// assert_eq!(rect.is_zero_sized(), false);
+    ///
+    /// let rect = Rectangle::new(Point::new(10, 20), Size::zero());
+    /// assert_eq!(rect.is_zero_sized(), true);
+    /// ```
+    // MSRV: Add const when upgrading to at least 1.46.0
+    pub fn is_zero_sized(&self) -> bool {
+        self.size.height == 0 || self.size.width == 0
+    }
+}
+
+/// Checks if the two ranges overlap.
+fn overlaps(first: RangeInclusive<i32>, second: RangeInclusive<i32>) -> bool {
+    second.contains(first.start())
+        || second.contains(first.end())
+        || first.start() < second.start() && first.end() > second.end()
+}
+
+/// Anchor point.
```

Comment:
Should we move this type into the `geometry` module? It might be reused for other primitives in the future.